proposals/NNNN-isolated-any-functions.md

```diff
@@ -0,0 +1,671 @@
+thereafter be treated as a `MainActor`-isolated function; the fact that
+it was originally a non-isolated function is both statically and
+dynamically erased.
+
+### Runtime behavior
+
+Calling a `@isolated(any)` function value behaves the same way as a direct
+call to a function with that isolation would:
+
+- If the function is `async`, it will run with its formal isolation.  This
+  includes leaving isolated contexts if the function is dynamically
+  non-isolated, as specified by [SE-0338](https://github.com/apple/swift-evolution/blob/main/proposals/0338-clarify-execution-non-actor-async.md).
+
+- If the function is synchronous, it will run with its formal isolation
+  only if it is dynamically isolated.  If it is dynamically non-isolated,
+  it will simply run synchronously in the current context, even if that
+  is isolated, just like an ordinary call to a non-isolated synchronous
+  function would.
+
+### `isolation` property
+
+Values of `@isolated(any)` function type have a special `isolation`
+property.  The property is read-only and has type `(any Actor)?`.  The
+value of the property is determined by the dynamic isolation of the
+function value:
+
+- If the function is dynamically non-isolated, the value of `isolation`
+  is `nil`.
+- If the function is dynamically isolated to a global actor type `G`,
+  the value of `isolation` is `G.shared`.
+- If the function is dynamically isolated to a specific actor reference,
+  the value of `isolation` is that actor reference.
+
+### Distributed actors
+
+Function values cannot generally be isolated to a distributed actor
+unless the actor is known to be local.  When a distributed actor *is*
+local, function values isolated to the actor can be converted to
+`@isolated(any)` type as above.  The `isolation` property presents
+the distributed actor as an `(any Actor)?` using the same mechanism
+as [`#isolation`](https://github.com/apple/swift-evolution/blob/main/proposals/0420-inheritance-of-actor-isolation.md#isolated-distributed-actors).
+
+### Isolation checking
+
+Since the isolation of a dynamically-isolated function value is
+statically unknown, calls to it typically cross an isolation barrier.
+This means that the call must be `await`ed even if the function is
+synchronous, and the arguments and result must satisfy the usual
+sendability restrictions for cross-isolation calls.
+
+If `f` is an immutable binding (such as a local `let` or a non-`inout`
+parameter) of `@isolated(any)` function type, then a call to `f` does
+not cross an isolation barrier if the current context is isolated
+to a derivation of the expression `f.isolation`.
+
+A context is isolated to a derivation of a expression `E` if it has an
+isolated capture that is initialized to a derivation of `E`.
+
+An expression is a derivation of `E` if:
+
+- it has the exact form of `E`;
+- it is a reference to a capture or immutable binding immediately
+  initialized with a derivation of `E`; or
```

Comment:
IIRC, the current isolation analysis for isolated parameters does not look through an "immutable binding immediately initialized with a derivation of `E`", e.g.,

```swift
actor A {
  func g() { }
}

func f(actor: isolated A) {
  let a = actor
  a.g() // error: a is not considered to be in the same isolation domain, call needs to be asynchronous
}
```

If we want to generalize that, great, but we should do so consistently rather than make it a rule specific to `.isolation` on an `@isolated(any)` type.